### PR TITLE
MM 7970: Implement desktop bridge to allow for user activity status updates to be sent to the webapp

### PR DESCRIPTION
**Pull request by deanwhillier**
#### Summary
Implements a bridge for the Desktop app to communicate with the Webapp. Adds a handler for responding to user activity updates to maintain a logged in users online status while they are using their computer, even when the Desktop app is in the background.

This PR is part of several related PR's:
Desktop: [PR 945](https://github.com/mattermost/desktop/pull/945)
Webapp: This PR
Redux: [PR 795](https://github.com/mattermost/mattermost-redux/pull/795)
Server: [PR 10469](https://github.com/mattermost/mattermost-server/pull/10469)

#### Ticket Link
[MM-7970](https://mattermost.atlassian.net/browse/MM-7970), supported by [MM-14582](https://mattermost.atlassian.net/browse/MM-14582)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [ ] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: utils/desktop_bridge.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+export function connect(callbackFunction) {
+    // store provided callback for later use
+    callback = callbackFunction;
+
+    // connect bridge methods
+    window.desktopBridge.updateUserActivityStatus = updateUserActivityStatus;
+}
+
+const updateUserActivityStatus = (userIsActive) => {
```
Comment (lieut-data) [inline utils/desktop_bridge.js]: In order to facilitate an easier transition in the future should this API change, can we pass an object here instead ala `({userIsActive})`? Then the desktop app could add new parameters in the future as needed without having to worry about the parameter order to this method.

File: utils/desktop_bridge.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    callback('updateUserActivityStatus', {userIsActive});
```
Comment (lieut-data) [inline utils/desktop_bridge.js]: I know we check in the desktop app if `updateUserActivityStatus` is `null` before calling it, implying that `callback` will have been set, but I suppose it's possible for `connect` to have been called with a `null` `callbackFunction`. 0/5 on doing anything about this.

File: components/needs_team/needs_team.jsx [hunk 3]
```diff
@@ -80,6 +82,21 @@ export default class NeedsTeam extends React.Component {
 
         const team = this.updateCurrentTeam(this.props);
 
+        DesktopBridge.connect((event, data) => {
```
Comment (lieut-data) [inline components/needs_team/needs_team.jsx]: Maybe destructure the `data` directly? `(event, {userIsActive})`?

File: components/needs_team/needs_team.jsx [hunk 4]
```diff
@@ -80,6 +82,21 @@ export default class NeedsTeam extends React.Component {
...
+            switch (event) {
+            case 'updateUserActivityStatus':
+                // ping the server if the desktop app reports the user is still active
+                if (this.props.currentUser && data.userIsActive === true) {
+                    this.userIsActive = true;
```
Comment (lieut-data) [inline components/needs_team/needs_team.jsx]: It's a bit unfortunate to have another `userIsActive` bit in the system. 0/5 on whether it needs to happen now, but I wonder if we could expose a new action that itself embodies the logic of checking the old state and deciding whether to call `viewChannel`, abstracting the "hack" into a higher level. Then presumably the existing code that monitors focus/blur on the window could hook into the same action.

File: utils/desktop_bridge.js (hunk 1 above)
Comment (deanwhillier) [reply utils/desktop_bridge.js]: Redid the bridge to mimick the existing functionality of Electron's ipcMain and ipcRenderer modules already used on the Desktop side of this feature. An arbitrary number of properties are now supported.

File: utils/desktop_bridge.js (hunk 2 above)
Comment (deanwhillier) [reply utils/desktop_bridge.js]: Redid the bridge ... this has changed.

File: components/needs_team/needs_team.jsx (hunk 3 above)
Comment (deanwhillier) [reply components/needs_team/needs_team.jsx]: Redid the bridge ... this has changed. The new setup supports an arbitrary number of parameters.

File: components/needs_team/needs_team.jsx (hunk 4 above)
Comment (deanwhillier) [reply components/needs_team/needs_team.jsx]: That is not really needed for the initial version of this functionality – has been removed. There will possibly be a next version with some enhancements that could look into something like this if needed.

**Comment by deanwhillier** [comment]
Thanks @lieut-data! The bridge has been rewritten to support 2 way communication and is patterned after the Electron ipcMain/ipcRenderer modules already in use on the desktop side of this feature.

ipcMain/ipcRenderer convert provided parameters to JSON and back when transferring back and forth. Not really sure if that is strictly needed here, but I added to be consistent. I guess it could be removed since ipcMain/ipcRenderer already handle that between the webview holding the webapp and the rest of the desktop app. Thoughts?

**Comment by deanwhillier** [comment]
Closing to submit new PR ... lots has happened since this original.

**Pull request closed without merging**